File: leetcode/1011-capacity-to-ship-packages-within-d-days/solution.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int lo = 0, hi = 0;
        for (int w : weights) {
            lo = max(lo, w);
            hi += w;
        }

        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (feasible(weights, days, mid)) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return lo;
    }

private:
    bool feasible(vector<int>& weights, int days, int cap) {
        int usedDays = 1, cur = 0;
        for (int w : weights) {
            if (cur + w > cap) {
                usedDays++;
                cur = w;
                if (usedDays > days) return false;
            } else {
                cur += w;
            }
        }
        return true;
    }
};

File: leetcode/1011-capacity-to-ship-packages-within-d-days/solution.py
from typing import List


class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        def feasible(cap: int) -> bool:
            used_days, cur = 1, 0
            for w in weights:
                if cur + w > cap:
                    used_days += 1
                    cur = w
                    if used_days > days:
                        return False
                else:
                    cur += w
            return True

        lo, hi = max(weights), sum(weights)
        while lo < hi:
            mid = (lo + hi) // 2
            if feasible(mid):
                hi = mid
            else:
                lo = mid + 1
        return lo
